marker_api/src/context.rs

```diff
@@ -112,30 +101,133 @@ impl<'ast> AstContext<'ast> {
+    ///
+    /// ## Example 1
+    ///
+    /// ```
+    /// # use marker_api::prelude::*;
+    /// # marker_api::declare_lint!(
+    /// #     /// Dummy
+    /// #     LINT,
+    /// #     Warn,
+    /// # );
+    /// # fn value_provider<'ast>(cx: &AstContext<'ast>, node: ExprKind<'ast>) {
+    ///     cx.emit_lint(LINT, node, "<lint message>");
+    /// # }
+    /// ```
+    ///
+    /// The code above will roughly generate the following error message:
+    ///
+    /// ```text
+    ///  warning: <lint message>                <-- The message that is set by this function
+    ///  --> path/file.rs:1:1
+    ///   |
+    /// 1 | node
+    ///   | ^^^^
+    ///   |
+    /// ```
+    ///
+    /// ## Example 2
+    ///
+    /// ```
+    /// # use marker_api::prelude::*;
+    /// # marker_api::declare_lint!(
+    /// #     /// Dummy
+    /// #     LINT,
+    /// #     Warn,
+    /// # );
+    /// # fn value_provider<'ast>(cx: &AstContext<'ast>, node: ExprKind<'ast>) {
+    ///     cx
+    ///         .emit_lint(LINT, node, "<lint message>")
+    ///         .help("<text>");
+    /// # }
+    /// ```
+    ///
+    /// The [`DiagnosticBuilder::help`] call will add a help message like this:
+    ///
+    /// ```text
+    ///  warning: <lint message>
+    ///  --> path/file.rs:1:1
+    ///   |
+    /// 1 | node
+    ///   | ^^^^
+    ///   |
+    ///   = help: <text>               <-- The added help message
+    /// ```
+    ///
+    /// ## Example 3
+    ///
+    /// Creating suggestions can impact the performance. The
+    /// [`DiagnosticBuilder::decorate`] function can be used, to only add more
+    /// context to the lint emission if it will actually be emitted. This will
+    /// save performance, when the lint is allowed at the [`EmissionNode`]
+    ///
+    /// ```
+    /// # use marker_api::prelude::*;
+    /// # marker_api::declare_lint!(
+    /// #     /// Dummy
+    /// #     LINT,
+    /// #     Warn,
+    /// # );
+    /// # fn value_provider<'ast>(cx: &AstContext<'ast>, node: ExprKind<'ast>) {
+    ///     cx.emit_lint(LINT, node, "<lint message>").decorate(|diag| {
+    ///         // This closure is only called, if the lint is enabled. Here you
+    ///         // can create a beautiful help message.
+    ///         diag.help("<text>");
+    ///     });
+    /// # }
+    /// ```
+    ///
+    /// This will create the same help message as in example 2, but it will be faster
+    /// if the lint is enabled. The emitted message would look like this:
+    /// ```text
+    ///  warning: <lint message>
+    ///  --> path/file.rs:1:1
+    ///   |
+    /// 1 | node
+    ///   | ^^^^
+    ///   |
+    ///   = help: <text>               <-- The added help message
+    /// ```
     #[allow(clippy::needless_pass_by_value)] // `&impl ToString`
-    pub fn emit_lint<F>(
+    pub fn emit_lint(
         &self,
         lint: &'static Lint,
-        node: impl Into<EmissionNode>,
+        node: impl EmissionNode<'ast>,
         msg: impl ToString,
```

Comment:
I've tried a bit around with `impl ToOwned<Owned=String>` but couldn't get it to behave as nicely as `impl ToString` does. Is there a trick I've missed? :sweat_smile:

Reply:
Use `impl Into<String>` instead. I'm not sure what you "nicety" you meant

Reply:
It's just my usual amount of typos, it was supposed to say nicely. I'll try this :D

Reply:
`Into<String>` seems to work well :D